apps/shell/app.js

```diff
@@ -98,6 +110,32 @@ wss.on('connection', function connection (ws, req) {
   });
 });
 
+server.on('upgrade', function upgrade(request, socket, head) {
+  var queryParams = new URLSearchParams(url.parse(request.url).search);;
+  var requestToken = 'notdefined';
+
+  if (queryParams) {
+    requestToken = queryParams.get('csrf') || 'notdefined';
+  }
+
+  if (!tokens.verify(secret, requestToken)) {
+    socket.write([
+      'HTTP/1.1 401 Unauthorized',
+      'Content-Type: text/html; charset=UTF-8',
+      'Content-Encoding: UTF-8',
+      'Connection: close',
+      'X-OOD-Failure-Reason: bad csrf token',
+    ].join('\n') + '\n\n');
+
+    socket.end();
```

Comment:
It looks like that the ws package does the same thing. If you look at https://github.com/websockets/ws/blob/021d4ef4e3c61c2804651c77a97e4608deebb29c/lib/websocket-server.js#L181-L256 you will see that `handleUpgrade` in several places calls `abortHandshake` in particular this block https://github.com/websockets/ws/blob/021d4ef4e3c61c2804651c77a97e4608deebb29c/lib/websocket-server.js#L230-L253. Then if you look at `abortHandshake` itself https://github.com/websockets/ws/blob/021d4ef4e3c61c2804651c77a97e4608deebb29c/lib/websocket-server.js#L384-L406 you see it does `socket.write` then `socket.destroy();`.

So you might be able to call `wss.abortHandshake(socket, 401, 'bad csrf token', {'X-OOD-Failure-Reason':'bad csrf token'})` in place of this block.

Otherwise instead of using `\n` use `\r\n` and instead of using `socket.end()` use `socket.destroy()`.

Reply:
I think that makes sense to use their API as it's likely known to be stable.

Reply:
Their API is private, so I just changed to CRLFs and destroy.